File: firebase/functions/src/pairing.ts
// Vincular una caja con QR y doble código.
//
//   VOXON POS (caja, anónima)                  VOXON 360 (celular del dueño o gerente)
//   startPairing → muestra el QR ─────────────► claimPairing(QR)        → código A (6 números)
//   confirmPairingOnDevice(código A) → código B ─► completePairing(código B) → caja vinculada
//
// El doble código prueba que quien tiene el celular está frente a la caja: con una foto del QR
// no basta. El token del QR y los códigos se guardan con hash en pairings/{id}/secrets, que
// ninguna app puede leer; cada paso admite 5 intentos y todo vence a los 10 minutos.
import { FieldValue, Timestamp } from "firebase-admin/firestore";
import { HttpsError, onCall } from "firebase-functions/v2/https";

import { cleanName, requireUid, stringParam } from "./common";
import { enrollInTransaction, requireDeviceRoom, requireInstanceAdmin } from "./devices";
import { db } from "./firestore";
import { randomDigits, randomToken, sha256Hex } from "./security";

const PAIRING_LIFETIME_MS = 10 * 60_000;
const MAX_ATTEMPTS = 5;

type Snapshot = FirebaseFirestore.DocumentSnapshot;
type Status = "waiting" | "claimed" | "device_confirmed" | "linked" | "failed" | "cancelled";

function pairingRef(pairingId: string): FirebaseFirestore.DocumentReference {
  if (!/^[A-Za-z0-9]{10,40}$/.test(pairingId)) {
    throw new HttpsError("invalid-argument", "Ese QR no es de VOXON POS");
  }
  return db.doc(`pairings/${pairingId}`);
}

function secretsRef(ref: FirebaseFirestore.DocumentReference): FirebaseFirestore.DocumentReference {
  return ref.collection("secrets").doc("codes");
}

function statusMessage(status: string): string {
  switch (status) {
    case "linked":
      return "Esta caja ya quedó vinculada";
    case "failed":
      return "Demasiados códigos incorrectos. Empieza de nuevo en la caja.";
    case "cancelled":
      return "Se canceló la vinculación. Empieza de nuevo en la caja.";
    default:
      return "Este paso no corresponde ahora. Revisa lo que dice la caja.";
  }
}

/** La vinculación existe, no venció y está en el paso esperado. */
function requireStep(pairing: Snapshot, expected: Status): void {
  if (!pairing.exists) {
    throw new HttpsError("not-found", "Esa vinculación no existe. Pide un QR nuevo en la caja.");
  }
  const expiresAt: Timestamp | undefined = pairing.get("expiresAt");
  if (!expiresAt || expiresAt.toMillis() < Date.now()) {
    throw new HttpsError("deadline-exceeded", "La vinculación venció. Pide un QR nuevo en la caja.");
  }
  const status = pairing.get("status");
  if (status !== expected) {
    throw new HttpsError("failed-precondition", statusMessage(status));
  }
}

/** Cuenta un código incorrecto; al quinto la vinculación se cae. */
async function rejectCode(pairing: Snapshot): Promise<never> {
  const attempts = Number(pairing.get("attempts") ?? 0) + 1;
  const failed = attempts >= MAX_ATTEMPTS;
  await pairing.ref.update({ attempts, ...(failed ? { status: "failed" } : {}) });
  throw new HttpsError(
    "permission-denied",
    failed ? "Demasiados códigos incorrectos. Empieza de nuevo en la caja." : "Código incorrecto",
  );
}

function cleanCode(value: unknown): string {
  return stringParam(value).replace(/\s/g, "");
}

/** La caja nueva pide un QR. */
export const startPairing = onCall({ invoker: "public" }, async (request) => {
  const uid = requireUid(request.auth);
  const platform = stringParam(request.data?.platform).slice(0, 20) || "desconocida";

  const ref = db.collection("pairings").doc();
  const qrToken = randomToken();
  const expiresAt = Timestamp.fromMillis(Date.now() + PAIRING_LIFETIME_MS);
  const batch = db.batch();
  batch.set(ref, {
    deviceUid: uid,
    platform,
    status: "waiting",
    attempts: 0,
    claimedBy: null,
    instanceName: null,
    deviceName: null,
    createdAt: FieldValue.serverTimestamp(),
    expiresAt,
  });
  batch.set(secretsRef(ref), { qrTokenHash: sha256Hex(qrToken) });
  await batch.commit();

  return {
    pairingId: ref.id,
    qrToken,
    qr: `voxon://pair?p=${ref.id}&t=${qrToken}`,
    expiresAt: expiresAt.toDate().toISOString(),
  };
});

/** El celular escanea el QR y elige a qué negocio va la caja. Devuelve el código A. */
export const claimPairing = onCall({ invoker: "public" }, async (request) => {
  const uid = requireUid(request.auth);
  const ref = pairingRef(stringParam(request.data?.pairingId));
  const qrToken = stringParam(request.data?.qrToken);
  const instanceId = stringParam(request.data?.instanceId);
  const name = cleanName(request.data?.name || "Caja", "el nombre de la caja", 40);

  const { instance } = await requireInstanceAdmin(instanceId, uid);
  requireDeviceRoom(await db.doc(`accounts/${instance.get("accountId")}`).get());

  const mobileCode = randomDigits(6);
  const expiresAt = await db.runTransaction(async (transaction) => {
    const [pairing, secrets] = await Promise.all([transaction.get(ref), transaction.get(secretsRef(ref))]);
    requireStep(pairing, "waiting");
    if (!qrToken || sha256Hex(qrToken) !== secrets.get("qrTokenHash")) {
      throw new HttpsError("permission-denied", "Ese QR no es válido");
    }
    transaction.update(ref, {
      status: "claimed",
      instanceId,
      accountId: instance.get("accountId"),
      instanceName: instance.get("name"),
      deviceName: name,
      claimedBy: uid,
      claimedAt: FieldValue.serverTimestamp(),
      attempts: 0,
    });
    transaction.set(secretsRef(ref), { mobileCodeHash: sha256Hex(mobileCode) }, { merge: true });
    return (pairing.get("expiresAt") as Timestamp).toDate().toISOString();
  });

  return { mobileCode, instanceName: instance.get("name"), deviceName: name, expiresAt };
});

/** La caja escribe el código A del celular. Devuelve el código B. */
export const confirmPairingOnDevice = onCall({ invoker: "public" }, async (request) => {
  const uid = requireUid(request.auth);
  const ref = pairingRef(stringParam(request.data?.pairingId));
  const code = cleanCode(request.data?.code);

  const [pairing, secrets] = await Promise.all([ref.get(), secretsRef(ref).get()]);
  requireStep(pairing, "claimed");
  if (pairing.get("deviceUid") !== uid) {
    throw new HttpsError("permission-denied", "Esta vinculación es de otra caja");
  }
  if (!code || sha256Hex(code) !== secrets.get("mobileCodeHash")) {
    await rejectCode(pairing);
  }

  const posCode = randomDigits(6);
  const batch = db.batch();
  batch.update(ref, { status: "device_confirmed", attempts: 0, deviceConfirmedAt: FieldValue.serverTimestamp() });
  batch.set(secretsRef(ref), { posCodeHash: sha256Hex(posCode) }, { merge: true });
  await batch.commit();
  return { posCode };
});

/** El celular escribe el código B de la caja: la caja queda vinculada. */
export const completePairing = onCall({ invoker: "public" }, async (request) => {
  const uid = requireUid(request.auth);
  const ref = pairingRef(stringParam(request.data?.pairingId));
  const code = cleanCode(request.data?.code);

  const [pairing, secrets] = await Promise.all([ref.get(), secretsRef(ref).get()]);
  requireStep(pairing, "device_confirmed");
  if (pairing.get("claimedBy") !== uid) {
    throw new HttpsError("permission-denied", "Solo quien escaneó el QR puede terminar la vinculación");
  }
  if (!code || sha256Hex(code) !== secrets.get("posCodeHash")) {
    await rejectCode(pairing);
  }
  // Los permisos pueden haber cambiado en estos minutos.
  await requireInstanceAdmin(pairing.get("instanceId"), uid);

  const enrolled = await db.runTransaction(async (transaction) => {
    const current = await transaction.get(ref);
    requireStep(current, "device_confirmed");
    const result = await enrollInTransaction(transaction, {
      uid: current.get("deviceUid"),
      instanceId: current.get("instanceId"),
      accountId: current.get("accountId"),
      name: current.get("deviceName"),
      platform: current.get("platform"),
      source: "qr",
    });
    transaction.update(ref, {
      status: "linked",
      linkedAt: FieldValue.serverTimestamp(),
      deviceNumber: result.deviceNumber,
    });
    transaction.delete(secretsRef(ref));
    return result;
  });
  return enrolled;
});

/** La caja o quien escaneó cancelan antes de terminar. */
export const cancelPairing = onCall({ invoker: "public" }, async (request) => {
  const uid = requireUid(request.auth);
  const ref = pairingRef(stringParam(request.data?.pairingId));
  const pairing = await ref.get();
  if (!pairing.exists) {
    return { ok: true };
  }
  if (pairing.get("deviceUid") !== uid && pairing.get("claimedBy") !== uid) {
    throw new HttpsError("permission-denied", "Esta vinculación no es tuya");
  }
  if (pairing.get("status") === "linked") {
    throw new HttpsError("failed-precondition", statusMessage("linked"));
  }
  await ref.update({ status: "cancelled", cancelledAt: FieldValue.serverTimestamp() });
  return { ok: true };
});

File: packages/voxon_data/lib/src/models/pairing.dart
import 'package:cloud_firestore/cloud_firestore.dart';

/// Pasos de la vinculación de una caja con QR y doble código.
enum PairingStatus {
  /// La caja muestra el QR y espera que lo escaneen.
  waiting,

  /// El celular escaneó: la caja debe escribir el código del celular.
  claimed,

  /// La caja confirmó: el celular debe escribir el código de la caja.
  deviceConfirmed,

  linked,
  failed,
  cancelled,
  expired;

  static PairingStatus parse(String? value) => switch (value) {
    'waiting' => waiting,
    'claimed' => claimed,
    'device_confirmed' => deviceConfirmed,
    'linked' => linked,
    'failed' => failed,
    'cancelled' => cancelled,
    _ => expired,
  };

  bool get isFinished =>
      this == linked || this == failed || this == cancelled || this == expired;
}

/// Lo que devuelve `startPairing`: el QR que muestra la caja.
final class PairingTicket {
  const PairingTicket({
    required this.pairingId,
    required this.qrToken,
    required this.qr,
    required this.expiresAt,
  });

  factory PairingTicket.fromMap(Map<String, dynamic> data) => PairingTicket(
    pairingId: data['pairingId'] as String,
    qrToken: data['qrToken'] as String,
    qr: data['qr'] as String,
    expiresAt: DateTime.parse(data['expiresAt'] as String),
  );

  final String pairingId;
  final String qrToken;

  /// Texto que va dentro del QR: `voxon://pair?p=…&t=…`.
  final String qr;
  final DateTime expiresAt;
}

/// Estado en vivo de una vinculación (documento pairings/{id}).
final class PairingState {
  const PairingState({
    required this.id,
    required this.status,
    this.instanceId,
    this.instanceName,
    this.deviceName,
    this.deviceNumber,
    this.expiresAt,
  });

  /// [now] permite marcar como vencida una vinculación sin terminar.
  factory PairingState.fromMap(
    String id,
    Map<String, dynamic>? data, {
    DateTime? now,
  }) {
    if (data == null) {
      return PairingState(id: id, status: PairingStatus.expired);
    }
    final expiresAt = readDate(data['expiresAt']);
    var status = PairingStatus.parse(data['status'] as String?);
    if (!status.isFinished &&
        expiresAt != null &&
        expiresAt.isBefore(now ?? DateTime.now())) {
      status = PairingStatus.expired;
    }
    return PairingState(
      id: id,
      status: status,
      instanceId: data['instanceId'] as String?,
      instanceName: data['instanceName'] as String?,
      deviceName: data['deviceName'] as String?,
      deviceNumber: (data['deviceNumber'] as num?)?.toInt(),
      expiresAt: expiresAt,
    );
  }

  final String id;
  final PairingStatus status;

  /// Negocio al que va la caja, desde que el celular escaneó.
  final String? instanceId;
  final String? instanceName;
  final String? deviceName;

  /// Número de la caja dentro del negocio, cuando ya quedó vinculada.
  final int? deviceNumber;
  final DateTime? expiresAt;
}

/// Lo que recibe el celular al escanear: el código A para escribir en la caja.
final class ClaimedPairing {
  const ClaimedPairing({
    required this.mobileCode,
    required this.instanceName,
    required this.deviceName,
    required this.expiresAt,
  });

  factory ClaimedPairing.fromMap(Map<String, dynamic> data) => ClaimedPairing(
    mobileCode: data['mobileCode'] as String,
    instanceName: data['instanceName'] as String,
    deviceName: data['deviceName'] as String,
    expiresAt: DateTime.parse(data['expiresAt'] as String),
  );

  final String mobileCode;
  final String instanceName;
  final String deviceName;
  final DateTime expiresAt;
}

/// Caja ya vinculada a una instancia.
final class LinkedDevice {
  const LinkedDevice({
    required this.instanceId,
    required this.instanceName,
    required this.mode,
    required this.deviceName,
    required this.deviceNumber,
  });

  factory LinkedDevice.fromMap(Map<String, dynamic> data) => LinkedDevice(
    instanceId: data['instanceId'] as String,
    instanceName: data['instanceName'] as String,
    mode: data['mode'] as String,
    deviceName: data['deviceName'] as String,
    deviceNumber: (data['deviceNumber'] as num).toInt(),
  );

  final String instanceId;
  final String instanceName;
  final String mode;
  final String deviceName;

  /// Número de la caja dentro del negocio (1, 2, 3…); prefija sus ventas.
  final int deviceNumber;
}

/// Código escrito para vincular una caja sin celular.
final class DeviceCode {
  const DeviceCode({
    required this.code,
    required this.expiresAt,
    required this.instanceName,
    required this.deviceName,
  });

  factory DeviceCode.fromMap(Map<String, dynamic> data) => DeviceCode(
    code: data['code'] as String,
    expiresAt: DateTime.parse(data['expiresAt'] as String),
    instanceName: data['instanceName'] as String,
    deviceName: data['deviceName'] as String,
  );

  final String code;
  final DateTime expiresAt;
  final String instanceName;
  final String deviceName;

  /// `ABCD-2345`, más fácil de dictar.
  String get formatted => '${code.substring(0, 4)}-${code.substring(4)}';
}

/// Fecha desde Firestore (Timestamp), texto ISO o DateTime.
DateTime? readDate(Object? value) => switch (value) {
  Timestamp() => value.toDate(),
  DateTime() => value,
  String() => DateTime.tryParse(value),
  _ => null,
};
